[스티치] API 테스트 / 문서자동화 미션 제출합니다.

lxxjn0 opened the pull request.

wmakerjun committed:
feat(index): change index page endpoint from admin to service

wmakerjun committed:
refactor(mypage): change my page endpoint from '/me' to '/mypage'

wmakerjun committed:
feat(join): add join page with ajax

wmakerjun committed:
feat(login): add login page with ajax

wmakerjun committed:
feat(mypage): add mypage with ajax

wmakerjun committed:
feat(api): update service api

wmakerjun committed:
feat(favorite): update favorite page for read and delete

wmakerjun committed:
refactor(servie): update promise into async/await

moonyoungCHAE committed:
refactor: admin page 변경

moonyoungCHAE committed:
feat: Bearer 인증 기능 추가 및 테스트, 깨지는 테스트 수정

moonyoungCHAE committed:
test: Member Controller create 테스트 및 문서화

moonyoungCHAE committed:
docs: 1단계 기능 구현 목록 작성

moonyoungCHAE committed:
test: Member findByEmail 테스트

moonyoungCHAE committed:
test: Member Update 테스트

moonyoungCHAE committed:
test: Member Delete 테스트

moonyoungCHAE committed:
test: MemberControllerTest 테스트 환경 변경

moonyoungCHAE committed:
test: LoginMemberController 에서 Token 테스트

leewnsdud committed:
refactor: MemberControllerTest 테스트 검증 방식 수정

moonyoungCHAE committed:
refactor: package 이동

moonyoungCHAE committed:
feat: MemberController 분리 및 인수테스트 작성

 - LoginMemberController / AdminMemberController로 분리
 - LoginMemberController 인수테스트 작성 및 해당 기능 구현

moonyoungCHAE committed:
docs: to do list 작성 (Login 멤버의 회원 관리)

moonyoungCHAE committed:
test: LoginMemberControllerTest 구현

moonyoungCHAE committed:
docs: to-do list 체크 및 추가

moonyoungCHAE committed:
refactor: Interceptor에 대한 단위테스트 작성

- Request에 token이 존재하지 않는 경우에 대한 케이스 테스트
- Request에 비정상적인 token이 존재하는 경우에 대한 케이스 테스트

moonyoungCHAE committed:
refactor: Interceptor에 대한 단위테스트 작성

- Request에 token이 존재하지 않는 경우에 대한 케이스 테스트
- Request에 비정상적인 token이 존재하는 경우에 대한 케이스 테스트

moonyoungCHAE committed:
feat: AdminMemberControllerTest 문서화

 - api-guide.adoc 수정

moonyoungCHAE committed:
feat: LoginMemberControllerTest 문서화

 - 패키지명 변경 doc -> docs

moonyoungCHAE committed:
refactor: PageController, MemberPageController 통합

moonyoungCHAE committed:
docs: front 연동 세부사항 추가

moonyoungCHAE committed:
feat: 로그인, 회원정보 수정, 탈퇴 - Front 연동

moonyoungCHAE committed:
docs: 추가 기능구현목록 작성

moonyoungCHAE committed:
docs: 2단계 기능 구현 목록 작성

moonyoungCHAE committed:
refactor: 메서드 이동 및 잘못된 URI 수정

moonyoungCHAE committed:
test: 즐겨찾기 관리에 대한 인수테스트 작성 (mock controller)

moonyoungCHAE committed:
refactor: MockMvc 한글 인코딩 설정 및 프린트 중복 제거

moonyoungCHAE committed:
feat: FavoriteController 구현 및 테스트

moonyoungCHAE committed:
feat: FavoriteService 구현 및 테스트

 패키지 이동 favorite -> member/favorite

moonyoungCHAE committed:
feat: 즐겨찾기 관리 기능 구현, DB 추가, 테스트

moonyoungCHAE committed:
feat: 즐겨찾기 API Front 연동

moonyoungCHAE committed:
feat: 즐겨찾기 API 문서화

hwookim committed:
docs: To-do List 수정

hwookim committed:
refactor: 패키지 이동

hwookim committed:
refactor: 불필요 메소드 삭제

hongsii commented on src/main/java/wooteco/subway/web/member/BearerAuthInterceptor.java:
```diff
@@ -0,0 +1,41 @@
+        }
+
+        String email = jwtTokenProvider.getSubject(token);
+        request.setAttribute("loginMemberEmail", email);
```
상수로 관리할 수 있지 않을까요?

hongsii commented on src/main/java/wooteco/subway/web/member/FavoriteController.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<Void> createFavorite(@LoginMember Member member, @RequestBody FavoriteRequest request) {
```
클라이언트뿐만 아니라, 서버에서도 유효한 요청 파라미터인지 검증하도록 바꿔보면 어떨까요?

lxxjn0 replied:
`FavoriteRequest`의 필드에 `@NotBlank` 어노테이션을 추가하여 검증을 진행하도록 수정하였습니다.
그리고 발생한 예외에 대해서는 AdviceController를 구현하여 처리하도록 하였습니다!!!

hongsii commented on src/main/java/wooteco/subway/web/member/LoginMemberController.java:
```diff
@@ -19,27 +14,33 @@ public LoginMemberController(MemberService memberService) {
         String token = memberService.createToken(param);
-        return ResponseEntity.ok().body(new TokenResponse(token, "bearer"));
+        return ResponseEntity.ok()
+                .body(new TokenResponse(token, "Bearer"));
```
선언해둔 상수를 사용할 수 있지 않을까요?

hongsii commented on src/test/java/wooteco/subway/docs/FavoriteDocumentation.java:
```diff
@@ -0,0 +1,48 @@
+public class FavoriteDocumentation {
+    public static RestDocumentationResultHandler createFavorite() {
+        return document("user/favorite/create",
+                requestHeaders(
+                        headerWithName("Authorization").description("This is token which is Bearer Type")
+                ),
```
반복되는 필드는 재사용해보면 어떨까요?

lxxjn0 replied:
반복되는 부분에 대하여 메서드 분리를 진행해보았습니다!!

hongsii commented on src/test/java/wooteco/subway/service/member/FavoriteServiceTest.java:
```diff
@@ -0,0 +1,75 @@
+
+@SpringBootTest
+@Sql("/truncate.sql")
+class FavoriteServiceTest {
```
예외 케이스의 테스트도 작성해보면 어떨까요?

lxxjn0 replied:
이전에 잘못된 `FavoriteRequest` 에 대한 예외 케이스는 controller 테스트로 작성해두었고, `FavoriteService` 에서는 존재하지 않는 Favorite을 삭제할 경우 전체 개수의 변화가 없음을 테스트해두었습니다.

hongsii commented on src/main/java/wooteco/subway/config/WebMvcConfig.java:
```diff
@@ -3,35 +3,25 @@
-        registry.addInterceptor(basicAuthInterceptor).addPathPatterns("/me/basic");
-        registry.addInterceptor(sessionInterceptor).addPathPatterns("/me/session");
-        registry.addInterceptor(bearerAuthInterceptor).addPathPatterns("/me/bearer");
+        registry.addInterceptor(bearerAuthInterceptor).addPathPatterns("/me");
+        registry.addInterceptor(bearerAuthInterceptor).addPathPatterns("/me/*");
```
`addPathPatterns(...)` 가 가변인자를 받기 때문에, 하나의 메서드 안에서 여러 path를 전달해도 좋을 것 같아요!

hongsii commented on src/main/java/wooteco/subway/web/member/FavoriteController.java:
```diff
@@ -0,0 +1,48 @@
+
+@RequestMapping("/me/favorites")
+@RestController
+public class FavoriteController {
```
컨트롤러에서 예외가 발생하면 클라이언트로 어떤 응답이 나가고 있을까요?
`@ControllerAdvice` 를 활용해 에러를 핸들링 해보면 어떨까요?

hongsii reviewed (changes requested):
안녕하세요 스티치!
전반적으로 요구사항 구현을 잘 해주셨네요! 👍 
몇 가지 피드백 드렸으니, 참고하셔서 반영해주세요!

leewnsdud committed:
docs : 1차 피드백 리팩토링 진행목록 작성

leewnsdud committed:
refactor : Transactional 어노테이션 추가 및 반복적인 값에 대한 상수화

- Service 각각의 메서드들에 Transactional 어노테이션 추가.
- 반복적으로 사용하는 값에 대한 상수화 처리.

leewnsdud committed:
refactor : FavoriteController의 request에 대한 유효성 검증 로직 구현

- FavoriteRequest의 유효성 검증 구현.
- ValidateExceptionAdvice를 구현하여 매개변수 예외 발생 처리.

leewnsdud committed:
refactor : FavoriteDocumentation의 중복 부분 처리

leewnsdud committed:
refactor : FavoriteService의 예외 케이스에 대한 테스트 작성

- delete시 존재하지 않는 Favorite에 대한 삭제 요청 테스트 작성.

leewnsdud committed:
refactor : addInterceptors 메서드 내부 중복 제거

lxxjn0 commented:
피드백 주신 부분에 대하여 수정해보았습니다.
상세한 피드백 남겨주셔서 항상 감사드립니다 :)

hongsii commented on src/main/java/wooteco/subway/service/member/MemberService.java:
```diff
@@ -9,43 +11,57 @@
+
+	@Transactional(readOnly = true)
+	public Member findMemberByEmail(String email) {
+		return memberRepository.findByEmail(email).orElseThrow(RuntimeException::new);
```
의미 없는 예외보다 구체적인 예외를 통해 발생한 원인을 알려줄 수 있도록 바꿔보면 어떨까요?

lxxjn0 replied:
`RuntimeException` 이 발생하는 부분을 확인하니 모두 `MemberService` 이었습니다.
예외를 구체화시키기 위해 `InvalidMemberException` 으로 수정하였습니다!!

hongsii commented on src/main/java/wooteco/subway/web/LoginMemberMethodArgumentResolver.java:
```diff
@@ -0,0 +1,45 @@
+	@Override
+	public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer,
+		NativeWebRequest webRequest, WebDataBinderFactory binderFactory) {
+		String email = (String)webRequest.getAttribute("loginMemberEmail", SCOPE_REQUEST);
```
여기도 선언한 상수를 재사용하는게 어떨까요?

hongsii commented on src/main/java/wooteco/subway/web/controller/LoginMemberController.java:
```diff
@@ -0,0 +1,70 @@
+@RestController
+public class LoginMemberController {
+
+	private static final String BEARER_TOKEN_TYPE = "Bearer";
```
상수는 여러 군데에 중복해서 선언하는 것 보다 하나를 공유해서 사용하는 게 어떨까요?

lxxjn0 replied:
위의 피드백과 함께 적용하였습니다.
공통된 상수는 하나만 선언하고 public으로 함께 사용가능하도록 수정하였습니다.

hongsii commented on src/test/java/wooteco/subway/BearerAuthInterceptorTest.java:
```diff
@@ -0,0 +1,56 @@
+
+	@Test
+	void preHandle_EmptyToken() {
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		MockHttpServletResponse response = new MockHttpServletResponse();
```
이 클래스를 테스트하기 위해선 항상 request와 response가 필요합니다. 필드로 선언해 테스트 시작 전 초기화하도록 바꿔보면 어떨까요?

lxxjn0 replied:
반복적인 작업이어서 필드 변수로 분리하고 @BeforeEach로 처리해주니 코드가 훨씬 간결해졌습니다 :)
추가적으로 나머지 테스트 코드들도 리팩토링 진행해보았습니다!!!

hongsii commented on src/main/java/wooteco/subway/config/ValidateExceptionAdvice.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class ValidateExceptionAdvice {
```
아래의 경우도 처리해보면 어떨까요?

1. 패스워드가 틀렸거나, 로직상 발생하는 에러 처리
2. 개발자가 예상치 못한 에러가 발생 시 에러 처리

lxxjn0 replied:
기본적으로 로직상에서 의도적으로 발생하는 예외에 대한 Handler를 모두 선언해 두었습니다.
- MethodArgumentNotValidException
- InvalidAuthenticationException
- InvalidMemberException

그리고 개발자가 예상치 못한 에러가 발생하는 경우를 생각해봤는데 우선 DB에서 발생하는 예외의 경우 개발자가 예상치 못한 에러라고 생각이 들어서 아래의 예외를 핸들링 했습니다.
- DbActionExecutionException

그 외로 예상치 못한 에러가 발생하는 경우가 잘 생각이 나지 않아서 우선 Exception을 핸들링하는 메서드를 만들어뒀는데 혹시 더 나은 방법이 존재할까요?

hongsii replied:
예외 처리는 아주 잘 해주셨습니다! 💯 
그리고 예외 처리 시 HTTP 상태 코드를 모두 400으로 처리하셨는데, 상황에 맞는 적절한 HTTP 상태 코드를 내려주도록 바꿔보면 어떨까요?
* 인증되지 않은 요청 -> ?
* 예상치 못한 에러 -> ?

그리고 `@ExceptionHandler` 어노테이션의 인자로 여러 예외를 전달할 수 있습니다. 처리가 동일한 경우에는 하나의 핸들러에서 처리해도 좋을 것 같아요~

hongsii reviewed (changes requested):
피드백 반영을 깔끔하게 잘 해주셨네요!
간단한 피드백 몇가지 드렸으니, 확인하셔서 반영해주세요!

leewnsdud committed:
docs : 2차 피드백 리팩토링 진행목록 작성

leewnsdud committed:
refactor : MemberService에서 발생하는 예외를 구체적으로 나타내는 InvalidMemberException 구현

- InvalidMemberException 구현.

leewnsdud committed:
refactor : 여러 클래스에서 공통적으로 사용되는 상수 공유 처리

- Production Code에서 공통적으로 사용되는 상수 public으로 수정.
- Test Code에서 공통적으로 사용되는 상수 public으로 수정.
- 공통 상수를 재선언하지 않고 기존의 상수 사용.

leewnsdud committed:
refactor : 테스트 코드 전체적인 리팩토링

- BearerAuthInterceptorTest의 반복적인 변수를 필드 변수로 분리 진행.
- 상수화 처리가 남은 부분들 확인 후 수정.

leewnsdud committed:
refactor : 세부적인 예외 상황에 따른 ExceptionHandler 메서드 구현

- InvalidAuthenticationException, InvalidMemberException, DbActionExecutionException, Exception 에 대한 핸들링 메서드 구현.

lxxjn0 commented:
남겨주신 피드백을 기준으로 코드들을 리팩토링 해보았습니다.

피드백을 반영하면서 개발자가 예상하지 못한 예외를 어떻게 처리하는지 방법을 잘 몰라서 기존에 남겨두신 피드백 아래에 코멘트로 남겨두었습니다.
상세한 리뷰 남겨주셔서 감사합니다 :)

hongsii reviewed (changes requested):
피드백 반영을 깔끔하게 해주셨네요! 💯 
간단한 피드백 드렸으니, 여기까지 반영해주세요~

leewnsdud committed:
refactor : 상황에 맞는 에러 메시지를 내려주도록 수정

- 토큰의 유효성 오류를 unauthorized 에러로 반환.
- 동일한 기능을 하는 InvalidMemberExceptionHandler와 DbActionExecutionExceptionHandler 통합.

lxxjn0 commented:
피드백 주신 부분 수정하였습니다.
예상치 못한 에러를 어떻게 처리하는게 옳을지 생각을 하다가 우선 서버에서 잘 모르는 예외에 의한 오류여서 500 에러를 발생시키고 원인을 콘솔에 출력하는 식으로 구현해보았습니다.

hongsii commented on src/main/java/wooteco/subway/config/ValidateExceptionAdvice.java:
```diff
@@ -31,28 +32,23 @@
+
 		return ResponseEntity
-			.badRequest()
+			.status(HttpStatus.INTERNAL_SERVER_ERROR)
```
피드백 반영 👍 
예상치 못한 에러의 경우 클라이언트의 잘못된 요청(4xx)일 수도 있지만, 대부분 서버에서 잘못 처리 혹은 예상하지 못한 로직이기 때문에 5xx에러로 클라리언트에게 알려주는 것이 적절합니다.

hongsii reviewed (approved):
피드백 반영을 잘 해주셨네요! 👍 
미션 진행하시느라 고생 많으셨습니다! 머지할게요 :)
